Add Windows Support

namrog84 opened the pull request:
# Summary

Add Windows support to react-native-localize #137 

* What issues does the pull request solve? Please tag them so that they will get automatically closed once the PR is merged
* What is the feature? (if applicable)
It adds support for Windows.

* How did you implement the solution?
By adding the appropriate changes to support react-native-windows.

* What areas of the library does it impact?
Only additive, it shouldn't impact any existing areas. 

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->
1. I had to make some unchecked in changes.
    * I had removed the `yarn pod-update` from package.json's post install since that doesn't work on windows. But I didn't want to break or modify any existing flows. 
2. I ran `yarn start-metro` 
3. I ran Build Solution debug x64 and then Start Debugging. 
   * Using recent version of Visual Studio 2019.
4. I then was able to take this screenshot.
   * https://i.imgur.com/nQPSVak.png


### What's required for testing (prerequisites)?
Windows machine with recent version of node(yarn) and visual studio 2019. 

### What are the steps to reproduce (after prerequisites)?
Please see above "Test Plan" for how to run example. 

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| Windows   |    ✅     |

## Checklist

<!-- Check completed item, when applicable, via: [X] -->

- [X] I have tested this on a device and a simulator
- [X] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`.   
    - I didn't see a changelog.md
- [ ] I updated the typed files (TS and Flow)
- [x] I added a sample use of the API in the example project (`example/App.js`)
  - I added windows example project


Note:
Missing support for event listener trigger of localizationDidChange. Someone from my team will investigate and PR that in future PR.

namrog84 commented on example/src/AsyncExample.js:
```diff
@@ -175,8 +175,10 @@ const styles = StyleSheet.create({
   },
   value: {
     textAlign: "left",
+    color: "black",
```
Since above the background is overridden from default to white.  It made the text not appear. 

Windows wants to default to black background and white text for me. I am not sure if that is because my windows is running on a 'dark theme'. I tried addressing it in the windows specific parts but was unsuccessful.

This seems to be work, and stay aligned with original design.

namrog84 commented on example/package.json:
```diff
@@ -26,6 +27,7 @@
     "@babel/runtime": "^7.12.5",
     "babel-loader": "^8.2.1",
     "metro-react-native-babel-preset": "^0.59.0",
+    "rimraf": "^3.0.2",
```
I can revert the rimraf change, if so desired. As there still exists a problem here. 
I could add a `clean:windows` in addition to `clean`  but sadly there is no way I could find to conditionally run postinstall or a different postinstall on windows, without resorting to a fully external script.js

yarn install will always fail because postinstall will fail on windows because of lack of support for `yarn pod-update`
So for anyone wanting to iterate/test on windows, that will need to be manually modified and not checked in.
Unless there is a better suggested way to not break existing workflows.

My main goal was to not break any existing flows at this time, and figured if anyone was more interested in a seamless windows dev loop, they can investigate in that future. We have been using an similiar implementation of this for over a year and not need any changes, so I figured windows dev loop is pretty low pri.

namrog84 commented on android/.project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
These keep getting generated for some reason and no .gitignore for them?
I will delete the files

asklar commented on README.md:
```diff
@@ -12,6 +12,7 @@ A toolbox for your React Native app localization.
 <br>
 [![Platform - Android](https://img.shields.io/badge/platform-Android-3ddc84.svg?style=flat&logo=android)](https://www.android.com)
 [![Platform - iOS](https://img.shields.io/badge/platform-iOS-000.svg?style=flat&logo=apple)](https://developer.apple.com/ios)
+[![Platform - Windows](https://img.shields.io/badge/platform-Windows-0067b8.svg?style=flat&logo=windows)](https://www.microsoft.com/en-us/windows)
```
```suggestion
[![Platform - Windows](https://img.shields.io/badge/platform-Windows-0067b8.svg?style=flat&logo=windows)](https://www.microsoft.com/windows)
```

asklar commented on README.md:
```diff
@@ -91,6 +92,12 @@ public class MainApplication extends Application implements ReactApplication {
 
+### Windows Support
+
+Because this RNW package targets React Native 0.63.0+, you probably won't need to link it manually. Otherwise if it's not the case, follow these additional instructions. You also need to manual link the module on Windows when using React Native Windows prior to 0.63:
```
```suggestion
Because this RNW package targets React Native 0.63.0+, you probably won't need to link it manually. Otherwise if it's not the case, follow these additional instructions. You also need to manually link the module on Windows when using React Native Windows prior to 0.63:
```

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,190 @@
+
+std::string RNLocalizeModule::getCountryCode(std::string locale)
+{
+    return locale.substr(locale.find("-") + 1, locale.length());
```
this will be wrong for some locales, e.g.:

Bosnian (Cyrillic, Bosnia and Herzegovina)  bs-Cyrl-BA

See https://docs.microsoft.com/en-us/windows/uwp/publish/supported-languages

I think you might want to use the actual API instead of parsing the locale code: https://docs.microsoft.com/en-us/uwp/api/windows.globalization.geographicregion?view=winrt-19041

namrog84 replied:
Switched to new API!

asklar commented on windows/RNLocalize/RNLocalizeModule.h:
```diff
@@ -0,0 +1,44 @@
+
+    const std::array<std::string, 6> USES_FAHRENHEIT{{"BS", "BZ", "KY", "PR", "PW", "US"}};
+    const std::array<std::string, 3> USES_IMPERIAL{{"LR", "MM", "US"}};
+    const std::array<std::string, 11> USES_RTL_LAYOUT{
```
this should come from the OS, not hardcoded

https://docs.microsoft.com/en-us/uwp/api/windows.globalization.languagelayoutdirection?view=winrt-19041

namrog84 replied:
Updated!

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,190 @@
+
+std::string RNLocalizeModule::getTemperatureUnit(std::string countryCode)
+{
+    return std::find(USES_FAHRENHEIT.begin(), USES_FAHRENHEIT.end(), countryCode) != USES_FAHRENHEIT.end() ? "fahrenheit" : "celsius";
```
y no kelvin! :P

namrog84 replied:
I was trying to stay consistent with rest of module and I couldn't find any country that uses kelvin as their standard temperature unit. :P

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,190 @@
+{
+    std::string calendar = to_string(winrt::Windows::Globalization::Calendar().GetCalendarSystem());
+
+    // Window's Calendar system can return any of these values:
```
```suggestion
    // The Windows calendar APIs can return any of these values:
```

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,190 @@
+
+NumberFormatSettings RNLocalizeModule::getNumberFormatSettings(std::string locale)
+{
+    LPCWSTR locale_l = std::wstring(locale.begin(), locale.end()).c_str();
```
don't use L types, just use PCWSTR or wchar_t*. (prefer std:: types when possible though)
(here and other places)

namrog84 replied:
Removed all places of windows macros

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,190 @@
+    LPCWSTR num_l = num_w.c_str();
+
+    LPWSTR numberFormatBuffer = new TCHAR[9];
+    GetNumberFormatEx(locale_l, NULL, num_l, NULL, numberFormatBuffer, 9);
```
nullptr instead of NULL, here and other places

namrog84 replied:
Fixed

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,190 @@
+    std::wstring num_w = std::wstring(num_s.begin(), num_s.end());
+    LPCWSTR num_l = num_w.c_str();
+
+    LPWSTR numberFormatBuffer = new TCHAR[9];
+    GetNumberFormatEx(locale_l, NULL, num_l, NULL, numberFormatBuffer, 9);
```
don't use TCHAR, use wchar_t directly
don't allocate if not needed
don't refer to the size of a buffer more than once as it's possible to get things out of sync, use ARRAYSIZE

```suggestion
    wchar_t numberFormatBuffer[9]={};
    GetNumberFormatEx(locale_l, nullptr, num_l, nullptr, numberFormatBuffer, ARRAYSIZE(numberFormatBuffer));
```

namrog84 replied:
Fixed

asklar commented on windows/RNLocalize/RNLocalizeModuleTypes.h:
```diff
@@ -0,0 +1,75 @@
+    std::string timeZone;
+
+    REACT_FIELD(uses24HourClock, L"uses24HourClock")
+    bool uses24HourClock;
```
default value?

namrog84 replied:
Fixed!

asklar commented on windows/RNLocalize/RNLocalizeModuleTypes.h:
```diff
@@ -0,0 +1,75 @@
+    bool uses24HourClock;
+
+    REACT_FIELD(usesMetricSystem, L"usesMetricSystem")
+    bool usesMetricSystem;
```
default value

namrog84 replied:
Fixed!

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,190 @@
+std::vector<winrt::hstring> RNLocalizeModule::getLocaleNames()
+{
+    std::vector<winrt::hstring> locales;
+    for (winrt::hstring locale : winrt::Windows::Globalization::ApplicationLanguages::Languages())
```
do these APIs need to be called from the UI thread? on newer RNW, modules run on the JS thread by default

namrog84 replied:
We've been doing it this way for over a year in our app without issue. And we have run into plenty of other UI thread contextual issues.  I don't see anything that indicates it needs or has to be.

namrog84 commented:
Will address all the suggested changes Monday or Tuesday.  Not around this weekend.

namrog84 commented:
@zoontek no rush or anything but just wanted to toss a FYI. 

I updated all PR suggested changes from @asklar and have tested example app project of what is checked in as of this comment. 
So I think it's ready for your review or whatever the next steps are at your convenience. 

Just let me know if there are any further suggested changes I can do or change.

Thanks! :thumbsup:

asklar commented on windows/RNLocalize/RNLocalizeModule.h:
```diff
@@ -0,0 +1,42 @@
+
+    const std::array<std::string, 6> USES_FAHRENHEIT{{"BS", "BZ", "KY", "PR", "PW", "US"}};
+    const std::array<std::string, 3> USES_IMPERIAL{{"LR", "MM", "US"}};
+};
```
nit these can be static

namrog84 replied:
Changed!

mck1117 replied:
oops, they can be static, but they need to be initialized in the cpp, not in the header

asklar reviewed (approved):
:shipit:

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,212 @@
+  // "TaiwanCalendar", "ThaiBuddhistCalendar", "UmAlQuraCalendar"
+  // In all cases they end in "Calendar" and RNL doesn't expect this, so we must truncate it.
+  calendar = calendar.substr(0, calendar.size() - 8);
+
```
```suggestion
  calendar = calendar.substr(0, calendar.size() - (ARRAYSIZE("Calendar") -1) );
```

namrog84 replied:
Good suggestion!  Fixed!

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,212 @@
+  std::string formattedString = winrt::to_string(numberFormatBuffer);
+
+  numberFormatSettings.decimalSeparator = formattedString.substr(5, 1);
+  numberFormatSettings.groupingSeparator = formattedString.substr(1, 1);
```
is this right? isn't the decimal separator at position 4?

namrog84 replied:
It is right!  In this example we have.
num1000(L"1000.00");

But then in the formattedString we have  "1,000.00"
because of the added comma from `GetNumberFormatEx`, it moved it over to 5, and not 4.

asklar replied:
good point, can you please add a comment :)

namrog84 replied:
Added comment describing situation

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,212 @@
+  std::wstring num_w = std::wstring(num_s.begin(), num_s.end());
+
+  const wchar_t *locale_l = std::wstring(locale.begin(), locale.end()).c_str();
+  const int numberFormatBufferLength = 9;
```
this is creating a temporary std::string, taking its const c_str, assigning it to locale_l, then destroying the temporary (and therefore the pointer is dangling now)

Make a persistent copy onto a local variable instead

namrog84 replied:
Derp. fixed.

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,212 @@
+  // Use GetTimeZoneInformation to get the hour bias from GMT and then create a formatted string.
+  TIME_ZONE_INFORMATION TimeZoneInfo;
+  GetTimeZoneInformation(&TimeZoneInfo);
+
```
check return value

namrog84 replied:
Checking return value

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,212 @@
+      timeZoneBuffer,
+      sizeof(timeZoneBuffer),
+      "GMT%s%02d:00",
+      hourBias < 0 ? "-" : "+",
```
timezones are usually given as offsets from UTC since GMT can go in/out of DST

namrog84 replied:
Looking at the GetTimeZoneInformation api info, this is actually UTC, not GMT.  So I believe it's just a typo in the string here. Fixing.

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,212 @@
+      sizeof(timeZoneBuffer),
+      "GMT%s%02d:00",
+      hourBias < 0 ? "-" : "+",
+      hourBias < 0 ? -hourBias : hourBias);
```
is there no js API in this module to return whether we are in DST?

namrog84 replied:
It doesn't appear to be, no.

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,212 @@
+  // Use GetNumberFormatEx to get number formatting based on the passed locale.
+  std::string num_s("1000.00");
+  std::wstring num_w = std::wstring(num_s.begin(), num_s.end());
+
```
no need to create two copies, you can just say:
std::wstring num(L"1000")

namrog84 replied:
Whoops! Fixed!

asklar commented on windows/RNLocalize/RNLocalizeModule.cpp:
```diff
@@ -0,0 +1,212 @@
+
+  numberFormatSettings.decimalSeparator = formattedString.substr(5, 1);
+  numberFormatSettings.groupingSeparator = formattedString.substr(1, 1);
+
```
not sure if there is a JS api for this here, but different locales have different rules for _where_ to put the grouping separator
for example in India you'd write "1,000" and "10,000", but "1,00,000" for 100000

namrog84 replied:
There isn't an API in this library as far as I can tell to indicate how that should be handled.

Might need to be hardcoded per language if that ever gets added.

And most of the more powerful friendly looking windows api's aren't actually available to us in this context .

So we had much more limited options for this :(

asklar reviewed (commented):
thanks for addressing the comments; I had signed off but then I saw a dangling pointer (please fix before merging) and I added a couple more comments

namrog84 commented:
> thanks for addressing the comments; I had signed off but then I saw a dangling pointer (please fix before merging) and I added a couple more comments

All good, I'll address all these soon! Thanks again for looking at it some more.  Ill make your suggested changes and I see a few other things myself I'd like to clean up a bit more. The code was quite old/unchanged for us, so it's great to try and get this in a much better state before merging.

zoontek commented:
@namrog84 @asklar I saw there's a lot of unresolved discussions. Is this PR ready to be reviewed?
As I don't own a windows computer and it is extremely painful for me to test it (virtual machine) I'd rather be sure that it is.

Thanks!

namrog84 commented:
It is not ready yet. I need to address a few more things.  Just some real life things came up for me and I haven't been able to resolve them yet. Hopefully should be able to before end of this week.  I will @ you when I am ready again.

namrog84 commented:
I have now fixed and/or responded to all of @asklar comments.  I have updated last few things and I believe am ready for another look now. 

I have no further plans for changes unless any one has more suggestions or thoughts.  We can wait a little bit in case anyone else wants to add anything before @zoontek gets a VM ready. 

If there is anything I can do to make it easier for you (e.g. take screenshots/video of anything specific). Just let me know.

asklar reviewed (approved):
Thanks!

zoontek commented on package.json:
```diff
@@ -23,13 +24,15 @@
     "format": "prettier '**/*.{js,json,md,ts,tsx}' --write",
-    "prepare": "bob build"
+    "prepare": "bob build",
+    "windows": "react-native run-windows"
```
Shouldn't this be in the `/example` app?

namrog84 replied:
It does sound reasonable, I think I'll just remove this.  I believe this was auto added by some script.  I never use these helper script on windows, since VS has such a much better flow without it.

zoontek commented:
LGTM 👍

I trust you guy since I don't own a Windows computer and I can't use virtualization anymore (Apple M1)
